docs/development.rst

```diff
@@ -61,9 +117,106 @@ pySMT only to msat and z3, independently of which other solvers are
+answer to these points and a bit of motivation for the theory. In this
+way we can discuss possible changes and ideas before you start working
+on the code.
 
+Code for a new Theory
+---------------------
 
-Distributing PySMT
+A good example of theory extension is represented by the BitVector
+theory. In case of doubt, look at how the BitVector case (bv) has been
+handled.
+
+Adding a Theory to the codebase is done by following these steps:
+
+1. Tests: Add a test file ```pysmt/test/test_<theory>.py```, to
+demonstrate the use for the theory (e.g., ```pysmt/test/test_bv.py```).
+
+2. Operators: Add the (basic) operators in ```pysmt/operators.py```,
+create a constant for each operator, and extend the relevant
+structures.
+
+3. Typing: Extend ```pysmt/typing.py``` to include the types (sorts)
+of the new theory.
+
+4. Walker: Extend ```pysmt/walkers/generic.py``` to include one
+```walk_``` function for each of the basic operators.
+
+5. FNode: Extend ```is_*``` methods in
+```pysmt/fnode.py:FNode```. This makes it possible to check the type
+of an expression, obtaining additional elements (e.g., width of a
+bitvector constant).
+
+5. Typechecker: Extend ```pysmt/type_checker.py:SimpleTypeChecker```
+to include type-checking rules.
+
+6. FormulaManager: Create constructor for all operators, including
+syntactic sugar, in ```pysmt/formula.py:FormulaManager```
+
+At this point you are able to build expressions in the new
+theory. This is a good time to start running your tests.
+
+8. Printers: Extend ```pysmt/printers.py:HRPrinter``` to be able to
+print expressions in the new theory (you might need to do this
+earlier, if you need to debug your tests output).
+
+7. Examples: Extend ```pysmt/test/examples.py``` with at least one
+example formula for each new operator defined in FormulaManager. These
+examples are used in many tests, and will help you identify parts of
+the system that still need to be extended (e.g., Simplifier).
+
+8. Theories and Logics: Extend ```pysmt/logics.py``` to include the
+new Theory and possible logics that derive from this Theory. In
+particular, define logics for each theory combination that makes
+sense.
+
+9. SMT-LIB: Extend ```pysmt/smtlib/parser.py:SmtLibParser``` and
+```pysmt/smtlib/printers.py``` to support the new
+operators.
+
+10. Shortcuts: All methods that were added in FormulaManager need to
+be available in ```pysmt/shortcuts.py```.
+
+At this point all pySMT tests should pass. This might require
+extending other walkers to support the new operators.
+
+10. Solver: Extend at least one solver to support the Logic. This is
+done by extending the associated Converter (e.g.,
+```pysmt/solvers/msat.py:MSatConverter```) and adding at least one
+logic to its ```LOGICS``` field. As a bare-minimum, this will require
+a way of converting solvers-constants back into pySMT constants
+(```Converter.back()```).
+
+Packaging and Distributing PySMT
 ==================
```

Comment:
When building this with Sphinx I get the following warning:

```
Packaging and Distributing PySMT
==================
/docs/development.rst:220: WARNING: Title underline too short.
```

The underlining needs to be as long as the title.